isDocument and isDocumentArray type guards

PabloSzx opened the pull request:
fixes #261

PabloSzx commented:
fixes #261

hasezoey commented:
@PabloSzx, sorry, but it seems like pull request templates dont really work, could you replace your current description with the "fixes #261" (otherwise as far as i know, it wont close the issue)

and could you give me an example on how this would fix it please?

PabloSzx commented:
A very simple example working on my project
![Code_-_Insiders_N1pamTi20O](https://user-images.githubusercontent.com/8672915/61849309-d0944100-ae7e-11e9-89d6-d15627c8cef9.png)

super-ienien commented:
I think you should rename `isModel` because the type of a populated field is not a Model. It's a `Document`.

PabloSzx commented:
renamed it to isDocument and added isDocumentArray for array of Refs

```typescript
export function isDocumentArray<T>(docs: Ref<T>[]): docs is T[] {
  return Array.isArray(docs) && docs.every(v => isDocument(v));
}
```


also added tests for both type guards, and therefore examples, something like this

```typescript
class UserRef extends Typegoose {
  @prop({ ref: UserRef, default: null })
  public master: Ref<UserRef>;

  @arrayProp({ itemsRef: UserRef, default: [] })
  public subAccounts: Ref<UserRef>[];

  @prop({ required: true })
  public name: string;
}
const UserRefModel = new UserRef().getModelForClass(UserRef, {
  schemaOptions: {
    collection: "UserRefs",
  },
});
(async () => {
  const UserMaster = await UserRefModel.create({
    name: "master",
  });
  const UserSub = await UserRefModel.create({
    name: "sub",
    master: UserMaster._id,
  });
  UserMaster.subAccounts.push(UserSub._id);
  await UserMaster.populate("subAccounts").execPopulate();
  if (isDocumentArray(UserMaster.subAccounts)) {
    expect(UserMaster.subAccounts).to.have.lengthOf(1);
    for (const doc of UserMaster.subAccounts) {
      expect(doc.name).to.be.equal("sub");
      expect(doc.name).to.not.be.equal("other");
    }
  } else {
    throw new Error("Error in isDocumentArray");
  }
  await UserSub.populate("master").execPopulate();
  if (isDocument(UserSub.master)) {
    expect(UserSub.master.name).to.be.equal("master");
    expect(UserSub.master.name).to.not.be.equal("other");
  } else {
    throw new Error("Error in isDocument");
  }
})();

```

hasezoey commented on test/models/userRefs.ts:
```diff
@@ -0,0 +1,18 @@
+
+export const UserRefModel = new UserRef().getModelForClass(UserRef, {
+    schemaOptions: {
+        collection: 'UserRefs'
```
why use a specific collection?

PabloSzx replied:
because at first the tests had problems with another model instance, but with with the class name changed , now it's fixed, i will clean this

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+        });
+        const UserSub = await UserRefModel.create({
+            name: 'sub',
+            master: UserMaster._id
```
```suggestion
            master: UserMaster
```
It works without `._id` and is more clean

PabloSzx replied:
in practice this almost never happens, you have the ref and you want to explicitly populate, it's this way in order to work as example too

PabloSzx replied:
to improve the test as example i will add the _id in another line

hasezoey replied:
in everything i so far used typegoose in, it worked so, so i want to ask now, in which cases does it *not* work?

PabloSzx replied:
Then you are simply not using populate and with type safety accessing it's properties, you are using the refs as parameters for findById and filling the documents manually, is another way of doing it, but much more verbose

hasezoey replied:
when making it without id, mongoose just greps the model's, so it is the same... (i tested it always with type safety, thats why i asked "in which case this does *not* work"(this = without `._id`)

hasezoey replied:
and as an example: 
look at [#204](https://github.com/szokodiakos/typegoose/issues/204#issuecomment-515482936)

with the addition:
```ts
const found = await CPModel.findById(cp.id).populate("car").populate("buyer").exec();
// console.log(found);

console.log("buyer", isDocument(found.buyer));
console.log("car", isDocument(found.car));
```
with the output of
```
buyer true
car true
```

and in the DB
```json
{
  "_id": "5d3b137acceba85ceffac79f",
  "buyer": "5d3b137acceba85ceffac79d",
  "car": "5d3b137acceba85ceffac79c",
  "referral": {
    "referralDiscount": 1,
    "referrer": "5d3b137acceba85ceffac79d"
  },
  "__v": 0
}
```

 so i dont really understand where it should *not* work

PabloSzx replied:
You are not accessing the car and buyer properties, which is the whole purpose of the type guards, and giving the whole document as parameter or field is just not how refs really work, in population mongoose looks for the _ID

PabloSzx replied:
Try to do
```typescript
console.log(found.buyer.name);
console.log(found.car.model);
```

PabloSzx replied:
> when making it without id, mongoose just greps the model's, so it is the same... (i tested it always with type safety, thats why i asked "in which case this does _not_ work"(this = without `._id`)

the fact that mongoose handles it automatically does not mean it is doing the `.populate` function, and no matter how are you working around it, you can't access the properties of the populated document field without a type guard, which is in fact a documented feature of typescript
 [union type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types)
 [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types)

> Ref\<T\>
> ```javascript
> export type Ref<T> = T | mongoose.Schema.Types.ObjectId;
> ```

hasezoey replied:
I think we are talking past each other, i mean that `{ buyer: user, car, referral: rfl }` is the same as `{ buyer: user.id, car: car.id, referral: rfl.id }`, because they give the same result in the DB, and thus not modifying how the populate and with that the type guard works, my question was just about the `.id` and not about the type guard, because the type guard would work so or so

PabloSzx replied:
ok, i understand, but to explain better the use case in the test i just prefer to use populate explicitly, that's it

hasezoey replied:
*resolved*
thanks for taking the time to discuss about it

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+            name: 'sub',
+            master: UserMaster._id
+        });
+        UserMaster.subAccounts.push(UserSub._id);
```
```suggestion
        UserMaster.subAccounts.push(UserSub);
```
It works without `._id` and is more clean

PabloSzx replied:
in practice this almost never happens, you have the ref and you want to explicitly populate, it's this way in order to work as example too

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+        });
+        const UserSub = await UserRefModel.create({
+            name: 'sub',
+            master: UserMaster._id
```
```suggestion
            master: UserMaster
```
It works without `._id` and is more clean

PabloSzx replied:
in practice this almost never happens, you have the ref and you want to explicitly populate, it's this way in order to work as example too

to improve the test as example i will add the _id in another line

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+                expect(doc.name).to.not.be.equal('other');
+            }
+        } else {
+            throw new Error('Error in isDocumentArray');
```
```suggestion
            throw new Error('"UserMaster.subAccounts" is not an Document Array!');
```

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+            expect(UserSub.master.name).to.be.equal('master');
+            expect(UserSub.master.name).to.not.be.equal('other');
+        } else {
+            throw new Error('Error in isDocument');
```
```suggestion
            throw new Error('"UserSub.master" is not an Document!');
```

hasezoey commented on test/models/userRefs.ts:
```diff
@@ -0,0 +1,18 @@
+    public subAccounts: Ref<UserRef>[];
+
+    @prop({ required: true })
+    public name: string;
```
```suggestion
    public name!: string;
```

hasezoey commented on test/models/userRefs.ts:
```diff
@@ -0,0 +1,18 @@
+
+export class UserRef extends Typegoose {
+    @prop({ ref: UserRef, default: null })
+    public master: Ref<UserRef>;
```
```suggestion
    public master?: Ref<UserRef>;
```

hasezoey commented on test/models/userRefs.ts:
```diff
@@ -0,0 +1,18 @@
+    public master: Ref<UserRef>;
+
+    @arrayProp({ itemsRef: UserRef, default: [] })
+    public subAccounts: Ref<UserRef>[];
```
```suggestion
    public subAccounts!: Ref<UserRef>[];
```

hasezoey commented on src/utils.ts:
```diff
@@ -86,3 +87,19 @@ export function getClassForDocument(document: mongoose.Document): any {
 }
+
+/**
+ * Type guard for a given document if it's field is already populated
```
```suggestion
 * Check if the given document is already populated
```

hasezoey commented on src/utils.ts:
```diff
@@ -86,3 +87,19 @@ export function getClassForDocument(document: mongoose.Document): any {
+}
+
+/**
+ * Type guard for a given array of documents if it's field is already populated
```
```suggestion
 * Check if the given array is already populated
```

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+            master: UserMaster._id
+        });
+        UserMaster.subAccounts.push(UserSub._id);
+        await UserMaster.populate('subAccounts').execPopulate();
```
```suggestion
        await UserMaster.populate('subAccounts').execPopulate();
        
```

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,43 @@
+            name: 'sub',
+            master: UserMaster._id
+        });
+        await UserSub.populate('master').execPopulate();
```
```suggestion
        await UserSub.populate('master').execPopulate();
        
```

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,87 @@
+        expect(doc.name).to.be.an('undefined');
+      }
+    } else {
+      throw new Error('"UserMaster.subAccounts" is populated!');
```
```suggestion
      throw new Error('"UserMaster.subAccounts" is populated where it should not!');
```

hasezoey commented on test/tests/typeguards.test.ts:
```diff
@@ -0,0 +1,87 @@
+      // @ts-ignore
+      expect(UserSub.master.name).to.be.an('undefined');
+    } else {
+      throw new Error('"UserSub.master" is populated!');
```
```suggestion
      throw new Error('"UserSub.master" is populated where it should not!');
```

PabloSzx commented:
You are not accessing the properties of car and buyer, that is the whole point of the type guards.

hasezoey commented on src/utils.ts:
```diff
@@ -86,3 +87,19 @@ export function getClassForDocument(document: mongoose.Document): any {
+ * Check if the given array is already populated
+ * @param docs The Array of Refs with uncertain type
+ */
+export function isDocumentArray<T>(docs: Ref<T>[]): docs is T[] {
```
```suggestion
export function isDocumentArray<T>(docs: Ref<T>[]): docs is InstanceType<T>[] {
```

hasezoey commented on src/utils.ts:
```diff
@@ -86,3 +87,19 @@ export function getClassForDocument(document: mongoose.Document): any {
+ * Check if the given document is already populated
+ * @param doc The Ref with uncertain type
+ */
+export function isDocument<T>(doc: Ref<T>): doc is T {
```
```suggestion
export function isDocument<T>(doc: Ref<T>): doc is InstanceType<T> {
```

PabloSzx commented:
> after i am testing more with this (remaking the tests for #351) i noticed it returns _just_ "is class" and not "is instance of class" but that is what these two functions are testing

makes sense, it was also not handling the documented feature of [deep-populate](https://mongoosejs.com/docs/populate.html#deep-populate) so i made tests for this feature too (*should handle recursive populations - multiple populates* and *should handle recursive populations - single populate*).

also i think it's better for them to be on their own file, specially since they now have more complexity, so i moved them to a new file **src/types.ts** (*i think **InstanceType** and **ModelType** should be moved there too, but it's just my opinion 🤷*).

hasezoey commented:
my personal note: i would recommend renaming the file to something like `typeguards.ts` instead of `types` to not be mistaken by typescripts `type` keyword (in the meaning of having "all" the types in that file)

PabloSzx commented:
another thing i'm thinking of is to rename the type guards to `isPopulated` and `isPopulatedArray` since they should be used on populations anyways, and those names are more descriptive 🤔

hasezoey commented:
...good question to rename them so, but they *just* verify that they are InstanceType's, and not specifically for populate :thinking:
